lib/experimental/auto-inserting-blocks.php

```diff
@@ -0,0 +1,246 @@
+ * via `gutenberg_serialize_blocks()`, we are able to run filters
+ * on the parsed blocks.
+ *
+ * @param WP_Block_Template[] $query_result Array of found block templates.
+ * @return WP_Block_Template[] Updated array of found block templates.
+ */
+function gutenberg_parse_and_serialize_block_templates( $query_result ) {
+	foreach ( $query_result as $block_template ) {
+		if ( 'custom' === $block_template->source ) {
+			continue;
+		}
+		$blocks                  = parse_blocks( $block_template->content );
+		$block_template->content = gutenberg_serialize_blocks( $blocks );
+	}
+
+	return $query_result;
+}
+add_filter( 'get_block_templates', 'gutenberg_parse_and_serialize_block_templates', 10, 1 );
+
+/**
+ * Filters the block template object after it has been (potentially) fetched from the theme file.
+ *
+ * By parsing a block template's content and then reserializing it
+ * via `gutenberg_serialize_blocks()`, we are able to run filters
+ * on the parsed blocks.
+ *
+ * @param WP_Block_Template|null $block_template The found block template, or null if there is none.
+ */
+function gutenberg_parse_and_serialize_blocks( $block_template ) {
+
+	$blocks                  = parse_blocks( $block_template->content );
+	$block_template->content = gutenberg_serialize_blocks( $blocks );
+
+	return $block_template;
+}
+add_filter( 'get_block_file_template', 'gutenberg_parse_and_serialize_blocks', 10, 1 );
+
+// Helper functions.
+// -----------------
+// The sole purpose of the following two functions (`gutenberg_serialize_block`
+// and `gutenberg_serialize_blocks`), which are otherwise copies of their unprefixed
+// counterparts (`serialize_block` and `serialize_blocks`) is to apply a filter
+// (also called `gutenberg_serialize_block`) as an entry point for modifications
+// to the parsed blocks.
+
+/**
+ * Filterable version of `serialize_block()`.
+ *
+ * This function is identical to `serialize_block()`, except that it applies
+ * the `gutenberg_serialize_block` filter to each block before it is serialized.
+ *
+ * @param array $block The block to be serialized.
+ * @return string The serialized block.
+ *
+ * @see serialize_block()
+ */
+function gutenberg_serialize_block( $block ) {
+	$block_content = '';
+
+	/**
+	 * Filters a parsed block before it is serialized.
+	 *
+	 * @param array $block The block to be serialized.
+	 */
+	$block = apply_filters( 'gutenberg_serialize_block', $block );
+
+	$index = 0;
+	foreach ( $block['innerContent'] as $chunk ) {
+		if ( is_string( $chunk ) ) {
+			$block_content .= $chunk;
+		} else { // Compare to WP_Block::render().
+			$inner_block    = $block['innerBlocks'][ $index++ ];
+			$block_content .= gutenberg_serialize_block( $inner_block );
+		}
+	}
+
+	if ( ! is_array( $block['attrs'] ) ) {
+		$block['attrs'] = array();
+	}
+
+	return get_comment_delimited_block_content(
+		$block['blockName'],
+		$block['attrs'],
+		$block_content
+	);
+}
```

Comment:
we spoke about this, but it seems reasonable to propose this change to Core itself.

Reply:
Yeah, the long-term plan for the code merge to Core is to port a number of things over, including the added filter in `serialize_block` or the [aforementioned](https://github.com/WordPress/gutenberg/pull/51449#discussion_r1279439166) registry for auto-inserting blocks.

I've scattered a `@todo` note or two across the code but nothing that amounts to a complete plan of action yet, since I wanted to let it steep in Gutenberg for a bit, as there's a chance that we'll still change a few things around (see e.g. https://github.com/WordPress/gutenberg/pull/52969) 🙂